src/goto-instrument/contracts/dynamic-frames/dfcc_utils.cpp

```diff
@@ -0,0 +1,577 @@
+{
+  auto &goto_functions = goto_model.goto_functions;
+  auto &symbol_table = goto_model.symbol_table;
+
+  auto old_function = goto_functions.function_map.find(function_id);
+  INVARIANT(
+    old_function != goto_functions.function_map.end(),
+    "dfcc_utilst::wrap_function: function to wrap " + id2string(function_id) +
+      " must exist in the program");
+
+  // Register the wrapped function under the new name
+  std::swap(
+    goto_functions.function_map[wrapped_function_id], old_function->second);
+
+  // Check that new name really exists
+  INVARIANT(
+    goto_functions.function_map.find(wrapped_function_id) !=
+      goto_functions.function_map.end(),
+    "dfcc_utilst::wrap_function: wrapped function " +
+      id2string(wrapped_function_id) +
+      " shall exist in function_map at this point");
+
+  // Remove previous entry
+  goto_functions.function_map.erase(old_function);
+
+  // Check that the slot is empty in function_map
+  INVARIANT(
+    goto_functions.function_map.find(function_id) ==
+      goto_functions.function_map.end(),
+    "dfcc_utilst::wrap_function: the original function " +
+      id2string(function_id) +
+      " shall not exist in function_map anymore at this point");
+
+  // Add new symbol for the wrapped function in the symbol table
+  source_locationt sl;
+  sl.set_file("wrapped functions for code contracts");
+  sl.set_line("0");
+  const symbolt *old_sym = symbol_table.lookup(function_id);
+  symbolt wrapped_sym;
+  wrapped_sym = *old_sym;
+  wrapped_sym.name = wrapped_function_id;
+  wrapped_sym.base_name = wrapped_function_id;
+  wrapped_sym.location = sl;
+  const auto wrapped_found = symbol_table.insert(std::move(wrapped_sym));
+  INVARIANT(
+    wrapped_found.second,
+    "dfcc_utilst::wrap_function: wrapped function symbol " +
+      id2string(wrapped_function_id) + " already exists in the symbol table");
+
+  // Re-insert a symbol for `function_id` which is now the wrapper function
+  symbolt wrapper_sym;
+  wrapper_sym = *old_sym;
+
+  std::function<const irep_idt(const irep_idt &)> trans_param =
+    [&](const irep_idt &old_param) {
+      return id2string(old_param) + "_wrapper";
+    };
+
+  // create new code_type with renamed parameters for the wrapper
+  const auto &old_code_type = to_code_type(old_sym->type);
+  code_typet::parameterst new_params;
+  clone_parameters(
+    old_code_type.parameters(),
+    wrapper_sym.mode,
+    wrapper_sym.module,
+    wrapper_sym.location,
+    // the naming scheme is `function_id::param` + `param_suffix`
+    trans_param,
+    function_id,
+    new_params);
+
+  code_typet new_code_type(new_params, old_code_type.return_type());
+
+  wrapper_sym.type = new_code_type;
+
+  // Remove original symbol from the symbol_table
+  if(goto_model.symbol_table.remove(function_id))
+  {
+    log.error() << "dfcc_utilst::wrap_function: could not remove "
+                << function_id << " from the symbol table" << messaget::eom;
+    exit(0);
```

Comment:
No exit

Reply:
If the ultimate vision is to make transformations scriptable one day we need to be throwing exceptions rather than exiting.

Reply:
fixed